### ISSUE TITLE
FP use_self requires generic_const_exprs

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#10756]
### Summary

.

### Lint Name

use_self

### Reproducer

I tried this code:

```rust
struct Foo;

impl Foo {
    fn foo<const N: usize>(&self) -> usize {
        let f = self;
        f.bar::<{
            let f = Foo;
            f.bar::<7>()
        }>() + N
    }

    const fn bar<const M: usize>(&self) -> usize {
        M
    }
}

fn main() {
    let f = Foo;

    assert_eq!(f.foo::<13>(), 20)
}
```

I saw this happen:

```
warning: unnecessary structure name repetition
 --> const-arg-in-const-arg.rs:7:21
  |
7 |             let f = Foo;
  |                     ^^^ help: use the applicable keyword: `Self`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self
  = note: requested on the command line with `-W clippy::use-self`
```

I expected to see this happen:
Suggestion does not compile wif const_generic_exprs feature is not enabled

````
error: generic parameters may not be used in const operations
 --> const-arg-in-const-arg.rs:7:21
  |
7 |             let f = Self;
  |                     ^^^^ cannot perform const operation using `Self`
  |
  = note: type parameters may not be used in const expressions
  = help: use `#![feature(generic_const_exprs)]` to allow generic const expressions

error: aborting due to previous error
````

### Version

```text
rustc 1.71.0-nightly (6d140d59b 2023-05-06)
binary: rustc
commit-hash: 6d140d59bb957903b02ddb993450509d8d92b448
commit-date: 2023-05-06
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.2
```

**Comment by xtqqczze** [comment]
Still an issue with Rust 1.92: https://godbolt.org/z/PeqKfKvbc
